### PR TITLE
TrainingArguments no longer sets `ACCELERATE_MIXED_PRECISION` environment variable

### LINKED ISSUE #29510: TrainingArguments shouldn't set ACCELERATE_MIXED_PRECISION env variable
### Feature request

When a mixed precision data type (such as `fp16`) is provided to `TrainingArguments`, it currently sets the `ACCELERATE_MIXED_PRECISION` ([cf](https://github.com/huggingface/transformers/blob/main/src/transformers/training_args.py#L1610)) environment variable to this mixed precision data type value.

Consequently, the `Trainer` class uses this mixed precision setting even when `TrainingArguments` are not explicitly provided as `args=training_args`.

The expected behavior should be that `Trainer` does not use mixed precision unless it's explicitly stated in the `TrainingArguments` provided as `args=training_args`.

CC: @muellerzr & @pacman100

### Motivation

I encountered this behavior while debugging a model that was producing `nan`s as output logits during training.

To troubleshoot this issue, I opted to stop providing `TrainingArguments` to `Trainer` in order to minimize potential sources of error but unfortunately that did not solved the problem.

After extensive investigation, this is how I discovered that `TrainingArguments` automatically sets `Trainer`'s mixed precision, even when not explicitly provided as an argument.

To prevent others from encountering the same debugging process, I advocate for mixed precision to be set only when explicitly required by the user by providing `TrainingArguments` to `Trainer`.

### Your contribution

I'm interested in opening a pull request to address this issue. This a good opportunity for me to learn more about the codebase.

### PR DESCRIPTION
**Pull request by Thytu**
# What does this PR do?

TrainingArguments no longer sets `ACCELERATE_MIXED_PRECISION` environment variable and instead relies on newly added `AcceleratorConfig.mixed_precision` attribute. This prevents cases where mixed precision was activated even when Training arguments wasn't provided to `Trainer`.

* AcceleratorConfig now contains a new attribute named `mixed_precision`

* TrainerArguments no longer sets `ACCELERATE_MIXED_PRECISION` environment variable

* TrainerArguments sets AcceleratorConfig.mixed_precision attribute according to received arguments

* TrainerArguments sets deep_speed mixed precision even when activated through ACCELERATE_USE_DEEPSPEED env variable

Fixes #29510

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr & @amyeroberts

### CODE REVIEW

File: src/transformers/trainer_pt_utils.py [hunk 1]
```diff
@@ -1209,6 +1209,11 @@ class AcceleratorConfig:
         },
     )
 
+    mixed_precision: str = field(
+        default="no",
+        metadata={"help": "Whether or not use mixed precision. Generally already set by TrainerArguments."},
```
Comment (muellerzr) [inline src/transformers/trainer_pt_utils.py:1214]:
```suggestion
        metadata={"help": "Whether or not use mixed precision and what type. Generally already set by TrainerArguments or during `accelerate launch`"},
```

**Review by muellerzr — COMMENTED** [review]
Thanks! Left some comments

**Comment by Thytu** [comment]
Hello @muellerzr, thanks for the review!

Sorry for the delay, been juggling a few different projects. I'll save some time this week to tackle your comments :)

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -1601,13 +1601,11 @@ def __post_init__(self):
                 # no need to assert on else
 
         # if training args is specified, it will override the one specified in the accelerate config
-        if self.half_precision_backend != "apex":
```
[the remark this answers is not in the document]
Comment (Thytu) [reply src/transformers/training_args.py:1604]: Here in my understanding we are not removing the `apex` backend, we are setting the `mixed_precision_dtype` variable whatever the selected backend is.

It was meant to prevent a crash in case in reach [that](https://github.com/huggingface/transformers/pull/29562/files#diff-bfceaff300c851b8e24fc50dc6638482abaec8f7d2a718e877c3828c166bcf79R1748) part of the code when the selected backend is apex (thus the variable wouldn't be initialized).

But I may be missing something here, let me know if you want it to be reverted (aside the `os.environ["ACCELERATE_MIXED_PRECISION"] = mixed_precision_dtype`) :)

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -1746,6 +1744,9 @@ def __post_init__(self):
                     self.accelerator_config = AcceleratorConfig(**self.accelerator_config)
                 else:
                     self.accelerator_config = AcceleratorConfig.from_json_file(self.accelerator_config)
+
+            self.accelerator_config.mixed_precision = mixed_precision_dtype
```
[the remark this answers is not in the document]
Comment (Thytu) [reply src/transformers/training_args.py]: You are totally write, should be fixed by da1c2e4fe13505c524889be51efb0cb20e7eaf93

**Comment by Thytu** [comment]
Updated the PR according to you comments and answered the one about apex.

**Comment by Thytu** [comment]
Oh 💩, I guess this git pull/merge wasn't a good idea 🤡

**Review by muellerzr — APPROVED** [review]
Thanks! This looks much better :)

File: src/transformers/trainer_pt_utils.py [hunk 4]
```diff
@@ -1210,6 +1210,11 @@ class AcceleratorConfig:
         },
     )
 
+    mixed_precision: str = field(
+        default="no",
+        metadata={"help": "Whether or not use mixed precision and what type. Generally already set by TrainerArguments or during `accelerate launch`"},
```
Comment (amyeroberts) [inline src/transformers/trainer_pt_utils.py:1215]: This should specify: 
- What it means if "no" is set 
- What takes precendence from AcceleratorConfig, trainer arguments or accelerate launch i.e. what if it's set here to fp16 and set to "no" in the trainer arguments?

File: src/transformers/training_args.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/training_args.py:1625]: If this was preventing a crash before, why is it not crashing now?

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -1770,6 +1768,18 @@ def __post_init__(self):
                     self.accelerator_config = AcceleratorConfig(**self.accelerator_config)
                 else:
                     self.accelerator_config = AcceleratorConfig.from_json_file(self.accelerator_config)
+
+            if self.accelerator_config.mixed_precision == "no":
```
Comment (amyeroberts) [inline src/transformers/training_args.py:1772]: @muellerzr What does "no" mean for setting mixed precision. Is it meant to be "unset" or "don't use mixed precision"?

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this. 

Some small questions to make sure I've understood the intended behaviour before approval

File: src/transformers/training_args.py (hunk 5 above)
Comment (muellerzr) [reply src/transformers/training_args.py:1772]: Don't use mixed precision

**Comment by muellerzr** [comment]
Gentle ping @Thytu 🤗
